### ISSUE TITLE
With big LAS files potential memory error

### ISSUE THREAD

**Issue by VelizarVESSELINOV** [kinverarity1/lasio#50]
With some of the big LAS files `numpy.loadtxt(StringIO(s))` is crashing due to not enough memory.

In the past I was using the solution described here http://stackoverflow.com/questions/8956832/python-out-of-memory-on-large-csv-file-numpy

I can generate a big las file, but depends on the memory available in my docker container can crash the parser.

This is an example even before during replace is crashing.

Line by line of chunk by chunk should be used to optimize the memory footprint.

Wrap exampled #1: the STRT is wrongly setup (above Everest altitude) in this example, but is not strange to have files with a lot of rows :)

```
~Well Information Block
STRT.FT           -178957.0000: START DEPTH
STOP.FT              8372.0000: STOP DEPTH
STEP.FT                 0.5000: STEP
~Curve Information Block
DEPT.FT            0 000 00 00: Depth
R01.IN                        : 
R02.IN                        : 
R03.IN                        : 
R04.IN                        : 
R05.IN                        : 
R06.IN                        : 
R07.IN                        : 
R08.IN                        : 
R09.IN                        : 
R10.IN                        : 
R11.IN                        : 
R12.IN                        : 
R13.IN                        : 
R14.IN                        : 
R15.IN                        : 
R16.IN                        : 
R17.IN                        : 
R18.IN                        : 
R19.IN                        : 
R20.IN                        : 
R21.IN                        : 
R22.IN                        : 
R23.IN                        : 
R24.IN                        : 
R25.IN                        : 
R26.IN                        : 
R27.IN                        : 
R28.IN                        : 
R29.IN                        : 
R30.IN                        : 
R31.IN                        : 
R32.IN                        : 
R33.IN                        : 
R34.IN                        : 
R35.IN                        : 
R36.IN                        : 
R37.IN                        : 
R38.IN                        : 
R39.IN                        : 
R40.IN                        : 
R41.IN                        : 
R42.IN                        : 
R43.IN                        : 
R44.IN                        : 
R45.IN                        : 
R46.IN                        : 
R47.IN                        : 
R48.IN                        : 
R49.IN                        : 
R50.IN                        : 
R51.IN                        : 
R52.IN                        : 
R53.IN                        : 
R54.IN                        : 
R55.IN                        : 
R56.IN                        : 
R57.IN                        : 
R58.IN                        : 
R59.IN                        : 
R60.IN                        : 
idmx.IN                       : 
idmn.IN                       : 
idav.IN                       : 
idmd.IN                       : 
DEVI.IN                       : 
MINR.IN                       :  
~Parameter Information Block
~A  Depth   
#    R01         R02         R03         R04         R05         R06     
#    R07         R08         R09         R10         R11         R12     
#    R13         R14         R15         R16         R17         R18     
#    R19         R20         R21         R22         R23         R24     
#    R25         R26         R27         R28         R29         R30     
#    R31         R32         R33         R34         R35         R36     
#    R37         R38         R39         R40         R41         R42     
#    R43         R44         R45         R46         R47         R48     
#    R49         R50         R51         R52         R53         R54     
#    R55         R56         R57         R58         R59         R60     
#    idmx        idmn        idav        idmd        DEVI        MINR    
#
-178957.0000
   -999.2500   -999.2500   -999.2500   -999.2500   -999.2500   -999.2500
   -999.2500   -999.2500   -999.2500   -999.2500   -999.2500   -999.2500
   -999.2500   -999.2500   -999.2500   -999.2500   -999.2500   -999.2500
   -999.2500   -999.2500   -999.2500   -999.2500   -999.2500   -999.2500
   -999.2500   -999.2500   -999.2500   -999.2500   -999.2500   -999.2500
   -999.2500   -999.2500   -999.2500   -999.2500   -999.2500   -999.2500
   -999.2500   -999.2500   -999.2500   -999.2500   -999.2500   -999.2500
   -999.2500   -999.2500   -999.2500   -999.2500   -999.2500   -999.2500
   -999.2500   -999.2500   -999.2500   -999.2500   -999.2500   -999.2500
   -999.2500   -999.2500   -999.2500   -999.2500   -999.2500   -999.2500
   -999.2500   -999.2500   -999.2500   -999.2500   -999.2500   -999.2500
-178956.5000
   -999.2500   -999.2500   -999.2500   -999.2500   -999.2500   -999.2500
   -999.2500   -999.2500   -999.2500   -999.2500   -999.2500   -999.2500
   -999.2500   -999.2500   -999.2500   -999.2500   -999.2500   -999.2500
   -999.2500   -999.2500   -999.2500   -999.2500   -999.2500   -999.2500
   -999.2500   -999.2500   -999.2500   -999.2500   -999.2500   -999.2500
   -999.2500   -999.2500   -999.2500   -999.2500   -999.2500   -999.2500
   -999.2500   -999.2500   -999.2500   -999.2500   -999.2500   -999.2500
   -999.2500   -999.2500   -999.2500   -999.2500   -999.2500   -999.2500
   -999.2500   -999.2500   -999.2500   -999.2500   -999.2500   -999.2500
   -999.2500   -999.2500   -999.2500   -999.2500   -999.2500   -999.2500
   -999.2500   -999.2500   -999.2500   -999.2500   -999.2500   -999.2500
...
```

Example 2:

```
~Well Information Block
STRT.FT           -178956.5000: START DEPTH
STOP.FT             11541.7500: STOP DEPTH
STEP.FT                 0.0500: STEP
~Curve Information Block
DEPT.FT            0 000 00 00: Depth
GR.GAPI                       : Gamma Ray
CCL.                          : Casing Collar Locator
AMP3FT.MV                     : AMP3FT Amplitude
TT3FT.USEC                    : 3FT Travel Time
AMPCAL.MV                     : AMPCAL Amplitude
AMP5FT.MV                     : AMP5FT Amplitude
~Parameter Information Block
PPT.USEC                0.0000: Predicted Pipe Time
CASEWGHT.LB/FT         29.0000: Casing Weight
MAXAMPL.MV             62.2000: Maximum Amplitude
MINAMPL.MV              3.3000: Minimum Amplitude
MINATTN.DB/FT           0.8000: Minimum Attenuation
CASEOD.IN               7.0000: Casing O.D.
~A  Depth        GR         CCL        AMP3FT      TT3FT       AMPCAL      AMP5FT   
-178956.5000   -999.2500   -999.2500   -999.2500   -999.2500   -999.2500   -999.2500
-178956.4500   -999.2500   -999.2500   -999.2500   -999.2500   -999.2500   -999.2500
...
```

Exception example:

```
CRITICAL - Exception Information
CRITICAL - Type: <class 'MemoryError'>
CRITICAL - Value: 
CRITICAL - Traceback:
    l = read(f, autodetect_encoding=True)
  File "/usr/local/lib/python3.4/site-packages/lasio/__init__.py", line 26, in read
    autodetect_encoding_chars=autodetect_encoding_chars)
  File "/usr/local/lib/python3.4/site-packages/lasio/las.py", line 156, in __init__
    autodetect_encoding_chars=autodetect_encoding_chars)
  File "/usr/local/lib/python3.4/site-packages/lasio/las.py", line 211, in read
    data = reader.read_data(len(self.curves))
  File "/usr/local/lib/python3.4/site-packages/lasio/las.py", line 517, in read_data
    s = re.sub(eol_chars, " ", s)
  File "/usr/local/lib/python3.4/re.py", line 179, in sub
    return _compile(pattern, flags).sub(repl, string, count)

```

Another exception example in numpy this time:

```
ERROR    - Failed to read wrapped data: MemoryError
Traceback (most recent call last):
  File "/usr/local/lib/python3.4/site-packages/lasio/las.py", line 519, in read_data
    arr = numpy.loadtxt(StringIO(s))
MemoryError

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
    l = read(f, autodetect_encoding=True)
  File "/usr/local/lib/python3.4/site-packages/lasio/__init__.py", line 26, in read
    autodetect_encoding_chars=autodetect_encoding_chars)
  File "/usr/local/lib/python3.4/site-packages/lasio/las.py", line 156, in __init__
    autodetect_encoding_chars=autodetect_encoding_chars)
  File "/usr/local/lib/python3.4/site-packages/lasio/las.py", line 211, in read
    data = reader.read_data(len(self.curves))
  File "/usr/local/lib/python3.4/site-packages/lasio/las.py", line 522, in read_data
    traceback.format_exc().splitlines()[-1]))
lasio.las.LASDataError: Failed to read wrapped data: MemoryError
```

**Comment by kinverarity1** [comment]
I am working on this now.

[Benchmark script](https://gist.github.com/kinverarity1/e24762c17da6a7643c829e6ba2168ecd) and [results](https://gist.github.com/kinverarity1/e24762c17da6a7643c829e6ba2168ecd#file-result-txt).

This is still not capturing the full memory usage (at least judging by Windows' Task Manager) but it's something for a start.

**Comment by kinverarity1** [comment]
This has hopefully been addressed (if not fully optimized) by PR #170
